### PR TITLE
MinGW fixes - adds examples that pass on newer MinGW builds

### PR DESCRIPTION
**Pull request by MSP-Greg**
Based on fixes used in ruby-loco, some used since last year

typo  rime => time

### CODE REVIEW

File: core/array/pack/l_spec.rb [hunk 1]
```diff
@@ -29,7 +29,7 @@
     it_behaves_like :array_pack_32bit_be, 'L>'
   end
 
-  guard -> { platform_is wordsize: 32 or platform_is :mingw32 } do
+  guard -> { platform_is wordsize: 32 } do
```
Comment (eregon) [inline core/array/pack/l_spec.rb]: Could you simplify to just `platform_is wordsize: 32 do`?

Also, this used to fail on AppVeyor's Ruby IIRC (that's the reason of these guards), is the wordsize seen as 32 bits on MinGW on 64-bit Windows now?

File: core/file/atime_spec.rb [hunk 2]
```diff
@@ -15,7 +15,7 @@
     File.atime(@file).should be_kind_of(Time)
   end
 
-  platform_is :linux do
+  guard -> { platform_is :linux or (platform_is :windows and ruby_version_is '2.5') } do
```
Comment (eregon) [inline core/file/atime_spec.rb]: Does `stat -c%x` work on Windows then?

File: core/file/utime_spec.rb [hunk 3]
```diff
@@ -16,30 +21,53 @@
...
     File.utime(@atime, @mtime, @file1, @file2)
-    File.atime(@file1).to_i.should be_close(@atime.to_i, 2)
-    File.mtime(@file1).to_i.should be_close(@mtime.to_i, 2)
-    File.atime(@file2).to_i.should be_close(@atime.to_i, 2)
-    File.mtime(@file2).to_i.should be_close(@mtime.to_i, 2)
+    if @time_is_float
+      File.atime(@file1).should be_close(@atime, 0.0001)
+      File.mtime(@file1).should be_close(@mtime, 0.0001)
+      File.atime(@file2).should be_close(@atime, 0.0001)
+      File.mtime(@file2).should be_close(@mtime, 0.0001)
```
Comment (eregon) [inline core/file/utime_spec.rb]: What's the reason for this change? To test more precision on Windows?
I think that would be better as a separate extra example/test, and let this test be general and platform-independent (I think some systems only provide second resolution for `atime`, but I'm not sure anymore).

File: core/array/pack/l_spec.rb (hunk 1 above)
Comment (MSP-Greg) [reply core/array/pack/l_spec.rb]: @eregon

Sorry about leaving the guard.  There's no Appveyor testing on MinGW 32 bit.

>  is the wordsize seen as 32 bits on MinGW on 64-bit Windows now?

Off hand, I don't know.  The MSYS2/RubyInstaller2 builds (>= Ruby 2.4) are built with much newer toolchains.  Appveyor is only testing >= Ruby 2.4 64 bit, but Travis is testing 2.3.  

I'll check them with 32 bit versions of Ruby 2.3 & 2.4, and let you know the results.  Should these remain for platforms other than Windows?

File: core/file/atime_spec.rb (hunk 2 above)
Comment (MSP-Greg) [reply core/file/atime_spec.rb]: Technically, no.  Does it work in the CI environment, yes.

I believe both test-all & test-spec have *nix utilities here and there.  Since spec requires build tools for the CAPI tests, the build tools are accessible.  One also needs some *nix utilities to build Ruby under vc (mswin).

File: core/file/utime_spec.rb (hunk 3 above)
Comment (MSP-Greg) [reply core/file/utime_spec.rb]: > What's the reason for this change? To test more precision on Windows?

The change happened a while ago, and I don't recall if it was due to a change in Windows, newer (better) build tools, or what (it did require a commit or two in trunk, 2.5 at the time).  If it accidentally reverted back, it might cause issues.  Hence, the tightened tests.

It's patch I've been using in ruby-loco for a **long time**, I finally decided to move many of my test patches into 'core', as Windows is now fully tested on mswin & mingw.

> I think some systems only provide second resolution for atime, but I'm not sure anymore.

If it passed on Travis, I thought all was ok.  Thoughts?

File: core/file/atime_spec.rb (hunk 2 above)
Comment (eregon) [reply core/file/atime_spec.rb]: OK, that's good enough then :)

File: core/file/utime_spec.rb (hunk 3 above)
Comment (eregon) [reply core/file/utime_spec.rb]: > If it passed on Travis, I thought all was ok. Thoughts?

It also needs to pass on all platforms of https://rubyci.org/
With the current code, the improved precision is only tested on Windows, so on the upside there are little risks.

File: core/array/pack/l_spec.rb (hunk 1 above)
Comment (eregon) [reply core/array/pack/l_spec.rb]: > Should these remain for platforms other than Windows?

Yes, the wordsize guards are needed, but indeed maybe we can drop the exception of mingw32, please verify. I think I originally had the guards to fix failures on AppVeyor with 32-bit Ruby.

Comment (MSP-Greg) [reply core/array/pack/l_spec.rb]: Forgive me, I'm confused.

> I think I originally had the guards to fix failures on AppVeyor with 32-bit Ruby.

Umm, 'or', not 'and'.  Hence, I assumed they needed the guard on non-Windows 32 bit builds.

I just ran core files as they are after this PR, and core/array/pack/l_spec.rb reported the same number of examples and expectations when tested on 2.4.5 32 bit and trunk 64 bit.

Ruby 2.3 32 bit is something I think we should forget.  23 failures on core.  I haven't tried 2.3 64 bit.

I'll remove the guard statements...

**Comment by eregon** [comment]
I finally merged this PR, as it now passes CI (AppVeyor had old Ruby versions).
@MSP-Greg Thanks!
